api/upload.py

```diff
@@ -174,20 +209,36 @@ def upload(self, strategy):
             if not user:
                 self.abort(403, 'Uploading requires login')
 
+        if strategy in ['label', 'uid', 'uid-match', 'reaper']:
+            strategy = strategy.replace('-', '')
+            strategy = getattr(Strategy, strategy)
+        else:
+            self.abort(500, 'strategy {} not implemented'.format(strategy))
+
         context = {'uid': self.uid if not self.superuser_request else None}
 
-        # TODO: what enum
-        if strategy == 'label':
-            strategy = Strategy.labelupload
-        elif strategy == 'uid':
-            strategy = Strategy.uidupload
-        elif strategy == 'uid-match':
-            strategy = Strategy.uidmatch
-        elif strategy == 'reaper':
-            strategy = Strategy.reaper
+        # Request for upload ticket
+        if self.get_param('ticket') == '':
+            return self._create_ticket()
+
+        # Check ticket id and skip permissions check if it clears
+        ticket_id = self.get_param('ticket')
+        if ticket_id:
+            ticket = self._check_ticket(ticket_id)
+            if not self.origin.get('id'):
+                # If we don't have an origin with this request, use the ticket's origin
```

Comment:
Same question as before - I don't think it's possible to have an empty origin with an authenticated request.